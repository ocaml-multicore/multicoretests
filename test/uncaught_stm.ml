(* Test of the behaviour of STM tests with uncaught exceptions *)

let always_fail () = failwith "unexpected"

module UncaughtExcConf : STM.Spec = struct
  open STM

  type sut = unit
  type state = unit
  type cmd = AlwaysFail of sut

  let show_cmd = function AlwaysFail () -> "AlwaysFail ()"
  let arb_cmd _ = QCheck.(make ~print:show_cmd (Gen.pure (AlwaysFail ())))
  let init_state = ()
  let next_state _ _ = ()
  let init_sut _ = ()
  let cleanup _ = ()
  let precond _ _ = true
  let run c s = match c with AlwaysFail () -> Res (unit, always_fail s)

  let postcond c _ r =
    match (c, r) with AlwaysFail (), Res ((Unit, _), _) -> true | _ -> false
end

module UE = STM_sequential.Make (UncaughtExcConf)

let _ =
  QCheck_base_runner.run_tests_main
    [
      UE.agree_test ~count:10 ~name:"STM test of uncaught exceptions";
      UE.neg_agree_test ~count:10 ~name:"neg STM test of uncaught exceptions";
    ]
