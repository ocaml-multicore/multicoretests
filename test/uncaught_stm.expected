random seed: 260395858
generated error fail pass / total     time test name

=== Error ======================================================================

Test STM test of uncaught exceptions errored on:

   AlwaysFail ()


exception Failure("unexpected") raised but not caught while running AlwaysFail ()


=== Error ======================================================================

Test neg STM test of uncaught exceptions errored on:

   AlwaysFail ()


exception Failure("unexpected") raised but not caught while running AlwaysFail ()

================================================================================
failure (0 tests failed, 2 tests errored, ran 2 tests)
